include/posix/sys/types.h

```diff
@@ -63,6 +63,19 @@ typedef u32_t clockid_t;
 typedef unsigned long timer_t;
 typedef unsigned long useconds_t;
 
+typedef u32_t pthread_rwlockattr_t;
+
+typedef struct pthread_rwlock_obj {
+	struct k_sem sem;
```

Comment:
Since semaphores are just counters, I wonder if you can't implement a rwlock using only two semaphores: one counting the number of readers, and one the number of writers.  It should simplify things a lot.  I'm not sure if you can implement all POSIX behavior, though.

Reply:
I had this approach in mind initially, 
k_sem_init(&read_lock, N, N); //N is number of threads.
k_sem_init(&WR_lock, 1, 1); //N is number of threads.

Acquiring RD Lock :- 
1. Check if WR lock is occupied or not, if not then acquire RD Lock. RD Lock is guaranteed to be available if WR Lock is acquired.

Acquiring WR Lock :-
Acquire WR Lock if available. 

Few points :- 
1. A thread can acquire a RW Lock as a reader more than once too. So we need to see what should be Limit for Read semaphore.

Reply:
This approach seems the way to go, yes.  I just don't understand "RD Lock is guaranteed to be available if WR Lock is acquired": if the write lock is taken, no read locks should be given.

The man page says this:
```
       The maximum number of simultaneous read locks that an  implementation  guarantees  can  be
       applied  to a read-write lock shall be implementation-defined. The pthread_rwlock_rdlock()
       function may fail if this maximum would be exceeded.
```

We can start with a maximum of N (and returning EAGAIN if that fails), and we would be compliant.  This simplifies the implementation on our side.  If we need to change later, we can fix it later.

Reply:
Sure I will put this implementation for review by EOD.

"Check if WR lock is occupied or not, if not then acquire RD Lock. RD Lock is guaranteed to be available if WR Lock is acquired." ==> This is with new approach which I was thinking of after your comments.  I will put this alternative approach today and we can decide one among these 2 approaches.